nanofi/examples/tail_file.c

```diff
@@ -0,0 +1,270 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+    *
+    *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+#include "api/nanofi.h"
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <sys/stat.h>
+
+typedef struct flow_file_records {
+    flow_file_record ** records;
+    uint64_t len;
+} flow_file_records;
+
+typedef struct file_offset {
+    int offset;
+} file_offset;
+
+struct flow_file_records * flowfiles = NULL;
+struct file_offset * fileoffset = NULL;
+struct nifi_instance * instance = NULL;
+
+void transmit_flow_files(nifi_instance * instance) {
+    if (flowfiles) {
+        for (int i = 0; i < flowfiles->len; ++i) {
+            if (flowfiles->records[i]) {
+                transmit_flowfile(flowfiles->records[i], instance);
+            }
+        }
+    }
+}
+
+char * get_string(char * begin, char * end) {
+
+    if (begin == NULL || (end != NULL && end < begin)) {
+        return NULL;
+    }
+
+    int len = 0;
+    char * buff = NULL;
+
+    if (begin == end) {
+        buff = (char *)malloc(2 * sizeof(char));
+        strncpy(buff, begin, 1);
+        buff[1] = '\0';
+        return buff;
+    }
+
+    if (end) {
+        len = end - begin;
+        buff = (char *)malloc((len+1) * sizeof(char));
+        strncpy(buff, begin, len);
+        buff[len] = '\0';
+    }
+    else {
+        len = strlen(begin);
+        buff = (char *)malloc((len+1) * sizeof(char));
+        strncpy(buff, begin, len);
+        buff[len] = '\0';
+    }
+    return buff;
+}
+
+void split_string(char * str, char delim, char ** collector, int * num_strings)
+{
+    if (!str) return;
+
+    while (*str == delim) str++;
+
+    char * p = strchr(str, delim);
+    if (p) {
+        char * s = get_string(str, p);
+        if (s == NULL) {
+            return;
+        }
+        collector[*num_strings] = s;
+        (*num_strings)++;
+        split_string(p+1, delim, collector, num_strings);
+    }
+    else {
+        char * s = get_string(str, NULL);
+        if (s == NULL) {
+            return;
+        }
+        collector[*num_strings] = s;
+        (*num_strings)++;
+    }
+}
+
+void set_offset(int offset) {
+    fileoffset->offset = offset;
+}
+
+int get_offset() {
+    return fileoffset->offset;
+}
+
+void free_all_strings(char ** strings, int num_strings) {
+    for (int i = 0; i < num_strings; ++i) {
+        if (strings[i]) free(strings[i]);
```

Comment:
Nitpicking here, but calling free with NULL is NOP (and therefore OK) according to the standard and works on all reasonable platforms.